consensus/polybft/consensus_runtime.go

```diff
@@ -158,23 +160,26 @@ func (c *consensusRuntime) AddLog(eventLog *ethgo.Log) { //nolint
 	return // TODO: Delete this when metrics is established. This is added just to trick linter.
 }
 
-// NotifyProposalInserted is an implementation of fsmNotify interface
-func (c *consensusRuntime) NotifyProposalInserted(b *StateBlock) {
-	lastHeader := b.Block.Header
-	if c.isEndOfEpoch(lastHeader.Number) {
+// OnBlockInserted is called whenever fsm or syncer inserts new block
+func (c *consensusRuntime) OnBlockInserted(block *types.Block) {
+	// after the block has been written we reset the txpool so that the old transactions are removed
+	c.config.txPool.ResetWithHeaders(block.Header)
+
+	if c.isEndOfEpoch(block.Header.Number) {
 		// reset the epoch. Internally it updates the parent block header.
-		if err := c.restartEpoch(lastHeader); err != nil {
+		if err := c.restartEpoch(block.Header); err != nil {
 			c.logger.Error("failed to restart epoch after block inserted", "err", err)
 		}
 	} else {
-		// inside the epoch, update last built block header
-		c.lastBuiltBlock = lastHeader
+		c.lock.Lock()
```

Comment:
Why do we need a lock around `lastBuiltBlock` if it's only changed in `restartEpoch` and on new block finalization? Those two functions should not be executed from two routines if I'm not mistaken.

Reply:
OnBlockInserted is called from two go routines. Syncer and ibft